internal/server/instance/drivers/driver_qemu.go

```diff
@@ -5821,6 +5827,64 @@ func (d *qemu) Rename(newName string, applyTemplateTrigger bool) error {
 	return nil
 }
 
+// Indirection to detachDisk.
+func qemuDetachDisk(s *state.State, id int) func(string) error {
+	return func(name string) error {
+		inst, err := instance.LoadByID(s, id)
+		if err != nil {
+			return err
+		}
+
+		qemuInst, ok := inst.(*qemu)
+		if !ok {
+			return fmt.Errorf("Couldn't assert QEMU object from interface")
+		}
+
+		return qemuInst.detachDisk(name)
+	}
+}
+
+// Detach a disk from the instance.
+func (d *qemu) detachDisk(name string) error {
+	diskName := strings.TrimPrefix(name, qemuDeviceIDPrefix)
+	return d.state.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		object, err := dbCluster.GetInstance(ctx, tx.Tx(), d.project.Name, d.name)
+		if err != nil {
+			return err
+		}
+
+		id := object.ID
+		devices, err := dbCluster.GetInstanceDevices(ctx, tx.Tx(), id)
+		if err != nil {
+			return err
+		}
+
+		device, ok := devices[diskName]
+		if !ok {
+			return fmt.Errorf("Device %s not found", diskName)
+		}
+
+		device.Config["attached"] = "false"
+
+		config, ok := d.expandedDevices[diskName]
+		if !ok {
+			return fmt.Errorf("Couldn't find device %q", diskName)
+		}
+
+		dev, err := d.deviceLoad(d, diskName, config)
+		if err != nil {
+			return err
+		}
+
+		err = d.deviceStop(dev, true, "")
+		if err != nil {
+			return err
+		}
```

Comment:
I suspect one of these, probably the latter is triggering its own DB transaction, for example to update volatile keys or something. This then causes you to be opening a transaction from within a transaction, with that nested transaction now getting blocked on its parent.

Try splitting this into two transactions, one for gathering the instance date, then make the deviceLoad/deviceStop out of the DB transaction, then open a new transaction for the UpdateInstanceDevices, that should fix it.